[2기-C 길근오] PostRepository 구현 & 테스트, 피드백 반영

hymn-fly opened the pull request:
## 👩‍💻 요구 사항과 구현 내용
 - PostRepository 구현
 - PostRepository 테스트코드

## ✅ 피드백 반영사항 
 - Fixture 이용
 - UuidUtils validation 로직을 User 도메인 안으로 이동
 - createdBy 부분 BasicEntity에서 제거, Post, User에서 따로 관리

## ✅ PR 포인트 & 궁금한 점
 - UUID validation 로직을 그냥 도메인(User) 안에 작성을 해도 괜찮을지요??

 - createdBy를 User에서는 그냥 아예 없앴고, Post에서는 연관관계로 매핑된 user를 createdBy의 대용으로 생각했습니다. 이렇게 하는게 조금더 객체지향에 가까운 것일까요??

Commit:
수정
 - BaseEntity에서 createdBy 제거
 - Post에서 user로 createdBy 대체
 - Uuid validation logic User 도메인 안으로 이동

Commit:
수정
 - user fixture 사용

Commit:
수정
 - entity에 ToString, EqualsAndHashCode 추가
 -

Commit:
추가
 - PostRepository, 테스트코드 추가

Commit:
추가
 - post로부터 user 얻는 테스트코드 추가

woogiekim commented on springboot/src/main/java/org/spring/notice/domain/post/Post.java:
```diff
@@ -31,7 +31,7 @@ public class Post extends BaseEntity {
     private String content;
 
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
```
케스케이드를 넣는 기준이 어떻게 되나요?

hymn-fly replied:
원래 post 저장하면서 user까지 저장하는걸 생각했었는데, 명시적이지 못했던거같습니다. 
그리고 CascadeType.persist일 때 , 여러 개의 post를 저장할 때 같은 user를 가지고 있으면 user를 계속 insert를 하게 되더라구요
그래서 cascade는 그냥 제거해버렸습니다

woogiekim commented on springboot/src/main/java/org/spring/notice/domain/post/Post.java:
```diff
@@ -31,7 +31,7 @@ public class Post extends BaseEntity {
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
     @JoinColumn(name="user_id", referencedColumnName = "id")
     private User user;
```
기왕 writer 라는 네이밍을 사용했으니 여기도 적용하는건 어떤가요? 일관성있게요

hymn-fly replied:
넵!

woogiekim commented on springboot/src/main/java/org/spring/notice/domain/user/User.java:
```diff
@@ -32,22 +31,22 @@ public class User extends BaseEntity {
     private String uuid;
```
이게 왜 필요한건지 사실 궁금하긴합니다

hymn-fly replied:
아 uuid를 쓰면 나중에 db를 바꾸거나 옮길때 도움이 된다는 이야기를 어디서 들었어서 한번 써봤습니다..!
사실 Long으로 하는게 일반적이긴 하겠죠

woogiekim replied:
이뤄지지않을 혹은 너무 먼 미래를 걱정한거 같습니다

해당 코드는 결국 기술부채가 될 가능성이 커보입니다

woogiekim commented on springboot/src/main/java/org/spring/notice/domain/user/User.java:
```diff
@@ -62,23 +61,23 @@ private User(String uuid, String name, int age, String hobby, String createdBy){
-        this.createdBy = createdBy;
-
-        return this;
+    private boolean isValidUuid(String uuid){
+        try{
+            UUID uuid1 = UUID.fromString(uuid);
+        }
+        catch (IllegalArgumentException exception){
+            return false;
+        }
+        return true;
     }
```
이건 좋은 선택은 아닌거같네요

제가 물어본 의도는

true false로 리턴해줘야 하는가? 입니다

유효하지않은 uuid 가 들어온거면 이거는 버그 아니면 공격 둘중하나 아닐까요?

woogiekim commented on springboot/src/test/java/org/spring/notice/domain/post/PostRepositoryTest.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Test
+    void findByTitleTest() {
+        List<Post> posts = postRepository.findByTitle("테스트1");
+
+        assertThat(posts).containsExactly(post1);
```
테스트가 빈약해요

이걸로 정말로 타이틀로 검색해서 포스팅을 가져왔다고 확신할수 있을까요?

woogiekim commented on springboot/src/test/java/org/spring/notice/domain/post/PostRepositoryTest.java:
```diff
@@ -0,0 +1,59 @@
+    @AfterEach
+    void tearDown(){
+        postRepository.deleteAll();
+    }
```
이거 할필요없이 테스트 클래스에 트랜잭셔널 붙이면 롤백이 됩니다.

woogiekim commented on springboot/src/test/java/org/spring/notice/domain/post/PostRepositoryTest.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Test
+    @Transactional
+    void getUserFromPost(){
+        // Given
+        List<Post> posts = postRepository.findByTitle("테스트1");
+
+        // When
+        User user = posts.get(0).getUser();
+
+        // Then
+        assertThat(user.getName()).isEqualTo(createUser().getName());
+        assertThat(user.getAge()).isEqualTo(createUser().getAge());
+        assertThat(user.getHobby()).isEqualTo(createUser().getHobby());
+    }
```
무엇을하고싶은건지 모르겠습니다

hymn-fly replied:
연관관계를 이용해서 post로부터 user를 가지고 오는것을 테스트하려고 했는데 생각해보니 당연한거라 테스트 안해도 될것같기도합니다. 아직은 테스트해야 할 기준이 명확하지가 않네요,,

woogiekim commented on springboot/src/test/java/org/spring/notice/domain/post/PostTest.java:
```diff
@@ -28,4 +30,13 @@ class PostTest {
+
+        assertThat(post.getTitle()).isEqualTo("테스트 제목");
+        assertThat(post.getContent()).isEqualTo("테스트 내용");
+        assertThat(post.getUser()).usingRecursiveComparison().isEqualTo(createUser());
```
using...

이 부분은 왜하는 건가요?

hymn-fly replied:
동일성 비교(hashcode비교) 뿐 아니라 안에 있는 필드들까지 다 비교하기 위해서 사용했습니다!
EqualsAndHashCode로  id 만을 사용하면, 필드들이 달라도 isEqualTo의 결과가 같게 나오기 때문에 저 메서드를 사용했습니다ㅎ

woogiekim commented on springboot/src/test/java/org/spring/notice/domain/user/UserTest.java:
```diff
@@ -56,4 +57,12 @@ class UserTest {
+        String id = UUID.randomUUID().toString();
+        User user = User.create(id, "테스트 유저", 10, "코딩");
+
+        assertThat(user.getUuid()).isEqualTo(id);
```
이것만 확인해봐도 될까요?

woogiekim reviewed (changes requested):
writer로 바꿔주신부분은 제가 생각하기에는 좋다고 봅니다.

테스트작성에 아쉬운 부분이 있는데
코멘트 확인해주시면 좋을거같습니다!

hymn-fly committed:
수정
- Fixture User 같은 uuid 갖도록 수정
- Post 도메인에서 user-> writer (변수이름수정)
- PostRepositoryTest 테스트 보강
- User EqualsAndHashCode annotation 추가

hymn-fly committed:
수정. PostRepositoryTest
- UserRepository 추가
- cascade 제거
- PostRepositoryTest에서 setup부분 수정(영속화문제)

hymn-fly committed:
추가
 - PostRepository page findAll 메서드
 - 해당 메서드 테스트 코드

hymn-fly committed:
수정
 - PostRepositoryTest pageFindAllTest () 수정

woogiekim commented on springboot/src/main/java/org/spring/notice/domain/user/User.java:
```diff
@@ -32,22 +29,26 @@ public class User extends BaseEntity {
-    private User(String uuid, String name, int age, String hobby, String createdBy){
-        checkArgument(isValidUuid(uuid), "유효하지 않은 UUID 입니다");
+    private User(String uuid, String name, int age, String hobby){
+        try{
+            UUID.fromString(uuid); // uuid validation
+        } catch(IllegalArgumentException ex){
+            throw new IllegalArgumentException("uuid 형식이 잘못되었습니다");
+        }
```
이거는 굳이라는 생각이 들어요

메시지 찍으려고 트라이 캐치로 잡아줘야하나?

uuid 만드는 주체는 서버인데...? 라는 생각이 들어오

woogiekim commented on springboot/src/test/java/org/spring/notice/domain/post/PostRepositoryTest.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @Test
+    void findByTitleTest() {
+        //Given
+        // When
+        List<Post> test1_posts = postRepository.findByTitle("테스트1");
+        List<Post> test2_posts = postRepository.findByTitle("테스트2");
+
+        // Then
+        assertThat(test1_posts).hasSize(2);
+        assertThat(test1_posts).containsExactlyInAnyOrder(post1, post4);
+        assertThat(test2_posts).hasSize(2);
+        assertThat(test2_posts).containsExactlyInAnyOrder(post2, post5);
```
이거는 검색을 위한것인가여?

그렇다면 페이징이 필요하지않나라는 생각이 듭니다

woogiekim reviewed (approved):
다음 진행을 위해서 어프로브 하겠습니다.

코멘트 확인해주시고 궁금한점은 DM주세요
